executor/src/witgen/global_constraints.rs

```diff
@@ -266,7 +266,21 @@ fn propagate_constraints<T: FieldElement>(
     (known_constraints, remove)
 }
 
-pub fn get_roots<T: FieldElement>(
+/// Get simple expression roots for a given expression.
+///
+/// This function recursively evaluates an algebraic expression and attempts
```

Comment:
This function is very complex, so its description has to be water-tight. It has to define what a root is and so on. I think a wording in the from of "if X is returned, it means the expression is equivalent to ..." would be helpful since it leaves little room for misunderstanding. Also, I don't think we need to explain _how_ the function works, just what it returns.